### PR TITLE
Consensus api refactor

### LINKED ISSUE #172: Disentangle API from `MinimintConsensus`
Currently the API task holds a reference to `MinimintConsensus` which thus has to be `Sync`. This is problematic for #79 and generally not a nice design imo. There should be some sort of accessor object that only requires access to the DB (which is already `Sync`).

### PR DESCRIPTION
**Pull request by jkitman**
Refactors a `ConsensusApi` out of `FedimintConsensus`.  Stops us mixing the API thread code and the consensus thread code, which also required some locks that could slow down the consensus thread unnecessarily.

Also cleans up a lot of the unnecessary complexity in building the `FedimintServer` + `FedimintConsensus` + `ConsensusApi`.

Will help us move to a final `FedimintApi` that combines the consensus and config gen APIs.

Fixes  #172

### CODE REVIEW

**Review by elsirion — DISMISSED** [review]
Overall great change, some details I want to make sure I understand correctly.

File: fedimint-server/src/consensus/mod.rs [hunk 1]
```diff
@@ -522,12 +391,6 @@ impl FedimintConsensus {
             let mut peers = maybe_exists.unwrap_or_default();
             peers.extend(upgrade_signals.iter().map(|(peer, _)| peer));
             dbtx.insert_entry(&ConsensusUpgradeKey, &peers).await;
-
-            // Remove our upgrade signal event if we signaled
-            if peers.contains(&self.cfg.local.identity) {
-                let mut cache = self.api_event_cache.lock().expect("locks");
-                cache.remove(&ApiEvent::UpgradeSignal);
-            }
```
[the remark this answers is not in the document]
Comment (jkitman) [reply fedimint-server/src/consensus/mod.rs:592]: Now that `FedimintServer` owns `FedimintConsensus` (no other threads could own it) we clear the cache in `process_events_then_propose` after flushing it to our proposal.

File: fedimint-server/src/lib.rs [hunk 2]
```diff
@@ -119,41 +123,30 @@ impl FedimintServer {
     pub async fn run(
         cfg: ServerConfig,
         db: Database,
-        module_gens: ServerModuleGenRegistry,
+        module_inits: ServerModuleGenRegistry,
```
[the remark this answers is not in the document]
Comment (jkitman) [reply fedimint-server/src/lib.rs:126]: Consistency with how it's named elsewhere, could change them all back.

File: fedimint-server/src/lib.rs [hunk 3]
```diff
@@ -459,12 +528,17 @@ impl FedimintServer {
...
+        &mut self,
+        override_proposal: Option<ConsensusProposal>,
+    ) -> ConsensusProposal {
         while let Some(Some(event)) = self.api_receiver.next().now_or_never() {
-            event_cache.insert(event);
+            self.consensus.api_event_cache.insert(event);
         }
+        let consensus_proposal = self.consensus.get_consensus_proposal().await;
+        self.consensus.api_event_cache.clear();
+        override_proposal.unwrap_or(consensus_proposal)
```
[the remark this answers is not in the document]
Comment (jkitman) [reply fedimint-server/src/lib.rs:541]: It's used only for testing to simulate a malicious server proposal everything else remaining the same, I wish there was a better way to inject it here.

File: fedimint-server/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dpc) [reply fedimint-server/src/lib.rs:126]: BTW. https://github.com/fedimint/fedimint/issues/2055

**Review by dpc — DISMISSED** [review]

**Pull request merged**